drivers/ethernet/eth_xlnx_gem.c

```diff
@@ -312,6 +312,37 @@ static void eth_xlnx_gem_isr(const struct device *dev)
 		}
 	}
 
+	/* DMA Rx buffer used bit not clear error handling */
+
+	if ((reg_val & ETH_XLNX_GEM_IXR_RX_USED_BIT) != 0x00000000U) {
+		sys_write32(ETH_XLNX_GEM_IXR_RX_USED_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IDR_OFFSET);
+		sys_write32(ETH_XLNX_GEM_IXR_RX_USED_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_ISR_OFFSET);
+		reg_val = sys_read32(dev_conf->base_addr +
+						ETH_XLNX_GEM_NWCTRL_OFFSET);
+		reg_val |= ETH_XLNX_GEM_NWCTRL_FLUSH_DPRAM_BIT;
+		sys_write32(reg_val, dev_conf->base_addr +
+						ETH_XLNX_GEM_NWCTRL_OFFSET);
+		struct eth_xlnx_gem_bd *bdptr;
+		uint32_t buf_iter;
+			bdptr = dev_data->rxbd_ring.first_bd;
+
+		for (buf_iter = 0; buf_iter < (dev_conf->rxbd_count - 1); buf_iter++) {
+			/* Clear 'used' bit -> BD is owned by the controller */
+			bdptr->ctrl = 0;
+			bdptr->addr = (uint32_t)dev_data->first_rx_buffer +
+			      (buf_iter * (uint32_t)dev_conf->rx_buffer_size);
+			++bdptr;
+		}
+		bdptr->ctrl = 0; /* BD is owned by the controller */
+		bdptr->addr = ((uint32_t)dev_data->first_rx_buffer +
+		      (buf_iter * (uint32_t)dev_conf->rx_buffer_size)) |
+		      ETH_XLNX_GEM_RXBD_WRAP_BIT;
```

Comment:
This doesn't make a whole lot of sense to me...

Let me try to recap: You've got a high traffic load situation. Flooding the transmit side of things is prevented by both the 'TX done' semaphore and the fact that the TX BD ring keeps track of how many buffers are currently available. Plus, the required CPU time to do so is probably spent somewhere else anyways. The RX side however is something we have little control over. Depending on the PHY link speed and the performance of your system, it can be reasonably expected that the controller will try to write more data to the buffers than they can hold. The controller detects this by looking at the current RX BD's 'used' bit, and if the buffer the controller wants to write to indicates that it's pending processing by the software, we get the RX used bit in the interrupt status register.

Things you could test to see if the frequency of those interrupts improves are:
- Try increasing the RX buffer count

Under normal circumstances, the hand-over of RX'ed packets from the buffers in the DMA to the network stack is an operation that is deferred to the system work queue. It is only once all currently available RX'ed packets have been processed that the receive interrupt gets re-enabled (the RX used interrupt may strike during that time, though). Therefore, you might also try:
- increasing the priority of the system work queue thread
- setting the device tree option that disables the deferred RX processing - upon a receive interrupt, the currently available packets will be forwarded within the context of the ISR. This is not advisable in general as it makes the time spent in the ISR pretty unpredictable under high load. It's definitely not deterministic behavior, but it cuts the context switch to the system work queue and doesn't make the operation susceptible to potentially being interrupted a lot.

It has to be expected though that under high loads, even with some increased resources and priority optimization, the controller will encounter situations where no buffers are available. The setting of the flush bit in the NWCTRL register seems to only tackle a very specific aspect of this problem, though. Here's what Xilinx commented in their bare-metal GEM driver:

> Fix for CR #692702 which updates error handling for Rx errors. Under heavy Rx traffic, there will be a large number of errors related to receive buffer not available. Because of a HW bug (SI #692601), under such heavy errors, the Rx data path can become unresponsive. To reduce the probabilities for hitting this HW bug, the SW writes to bit 18 to flush a packet from Rx DPRAM immediately. The changes for it are done in the function XEmacPs_IntrHandler.

So it seems that this operation is only meant to prevent the RX pipeline from stalling due to a hardware bug when no buffers are available by discarding the next pending packet from the pipeline at the stage *before* it is written to the DMA buffers.

Considering this, when the RX used interrupt is encountered, this doesn't necessarily mean that there's invalid packet data contained in the yet unprocessed RX buffers in the DMA or that RX processing as a whole is stuck. The deferred RX handler in the system work queue simply might not be able to keep up. Therefore, I don't see the need to invalidate all currently available data and clearing it for being overwritten by going over the entire RX BD ring and releasing the descriptors. Aside from this: once the DMAs, both RX and TX, have been set up, there should never be a need to re-assign each descriptor's address word. The control and address words are distinct 32-bit words and clearing the control words doesn't compromise the address words.

There's also another risk here that you have to consider: you're handling the RX used bit *after* the RX bit. At least if the GEM's interrupt is temporarily disabled at the GIC level or the global interrupt disable is active for some time, the GEM might first indicate a successful RX for buffer <n> and might encounter RX used in buffer <n+1>. The controller's interrupt line becomes re-enabled an now both status bits are pending in the ISR. Under normal circumstances (RX handling deferred to the work queue), first, the RX handling is enqueued, and then, this loop invalidates everything that there is to receive. This breaks the receive routine once it gets executed: the missing 'used' bit is detected and BD iteration is aborted, but now, the next buffer the controller is going to write to however likely won't match dev_data->rxbd_ring.next_to_process, likely stalling RX processing until the controller next writes to the buffer the local variable points to.

So if you really, really want a full purge (although I doubt that that's the best handling of the controller getting swamped in the RX path), the clean solution would be:
- disable reception completely in NWCTRL.
- Iterate the RX BDs and clear (only) the control word (minus the wrap bit in the last BD!).
- Set the RXQBASE register to the address of the first RX BD
- Set dev_data->rxbd_ring.next_to_process to 0.

If you have the impression that your RX side stalls completely under high load, and it's not just a massive processing backlog, then there might be one more aspect to look at which might stall RX processing on the driver side. In a local work-in-progress version of the driver yet to be committed (i.e., sometime after #74374), I once added this bit of code to the RX processing routine eth_xlnx_gem_handle_rx_pending, slightly modifying the original handling of a missing 'start of frame' bit:

```
if ((reg_val & ETH_XLNX_GEM_RXBD_START_OF_FRAME_BIT) == 0) {
	/*
	 * Although the current BD is marked as 'used', it doesn't
	 * contain the SOF bit.
	 *
	 * -> Unless we move the current RX BD pointer forward,
	 * we'll be stuck on this invalid entry - at least until
	 * the RX BDs wrap around to the next use of the current
	 * RX BD entry. Clear the 'used' bit before moving on.
	 */
	LOG_ERR("%s unexpected missing SOF bit in RX BD [%u]",
		dev->name, first_bd_idx);

	reg_val = sys_read32(reg_addr);
	reg_val &= ~ETH_XLNX_GEM_RXBD_USED_BIT;
	sys_write32(reg_val, reg_addr);

	dev_data->rxbd_ring.next_to_process =
		(dev_data->rxbd_ring.next_to_process + 1) %
		dev_conf->rxbd_count;

	continue;
}
```
The loop this handling is in will assure that any BD will be invalidated that has no SOF until an empty BD or one with a SOF bit is encountered, with a maximum of one full loop of the RX BDs. A missing SOF bit would actually point to data corruption and justifies BD invalidation even if there might be some (likely incomplete) data.

Maybe the possible RX processing backlog might also be reduced by removing the limit of one full iteration of the RX BD ring maximum with each execution of eth_xlnx_gem_handle_rx_pending. This is a safeguard to ensure that the system workqueue doesn't get blocked by continuous inbound packets. The hardware is likely faster than the software, so the RX routine could keep looping until the hardware actually gives it a break. The RX interrupt only becomes re-enabled once the function returns, so it's possible to handle a larger number of packets without more interrupts, whether this is advisable is another question, though.

Long story short: the RX used interrupts have to be expected during high load situations. Writing the flush bit is a Xilinx-approved safeguard against a potential hardware bug I didn't know about yet - I'm all for it. However, doing a full buffer purge of what's likely perfectly fine packet data, probably causing error handling by the peer like TCP re-transmits, doesn't seem to be a good solution. I'll try to reproduce the issue myself and check if there's actually a stall or just massive backlog, although my current hardware limits me to 100 MBps. I'd suggest that you try some more testing with modified values for the settings I've written about, and enable logging and check if there's ever a missing SOF bit in a high load situation.

There's likely a better solution for this.